src/plugins/discover/public/application/components/source_viewer/source_viewer.tsx

```diff
@@ -0,0 +1,113 @@
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+import React, { useEffect, useState } from 'react';
+import { monaco } from '@kbn/monaco';
+import { EuiEmptyPrompt, EuiLoadingSpinner } from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
+import { useEsDocSearch } from '../doc/use_es_doc_search';
+import { JsonCodeEditorCommon } from '../json_code_editor/json_code_editor_common';
+import { ElasticRequestState } from '../doc/elastic_request_state';
+import { getServices } from '../../../../public/kibana_services';
+import { SEARCH_FIELDS_FROM_SOURCE } from '../../../../common';
+
+interface SourceViewerProps {
+  id: string;
+  index: string;
+  indexPatternId: string;
+  hasLineNumbers: boolean;
+  width?: number;
+}
+
+export const SourceViewer = ({
+  id,
+  index,
+  indexPatternId,
+  width,
+  hasLineNumbers,
+}: SourceViewerProps) => {
+  const [editor, setEditor] = useState<monaco.editor.IStandaloneCodeEditor>();
+  const [jsonValue, setJsonValue] = useState<string>('');
+  const indexPatternService = getServices().data.indexPatterns;
+  const useNewFieldsApi = getServices().uiSettings.get(SEARCH_FIELDS_FROM_SOURCE);
+  const [reqState, hit] = useEsDocSearch({
+    id,
+    index,
+    indexPatternId,
+    indexPatternService,
+    requestAllFields: useNewFieldsApi,
+  });
+
+  useEffect(() => {
+    if (reqState === ElasticRequestState.Found) {
+      setJsonValue(JSON.stringify(hit, undefined, 2));
+    }
+  }, [reqState, hit]);
+
+  // setting editor height based on lines height and count to stretch and fit its content
+  useEffect(() => {
+    if (!editor) {
+      return;
+    }
+    const editorElement = editor.getDomNode();
+
+    if (!editorElement) {
+      return;
+    }
+
+    const lineHeight = editor.getOption(monaco.editor.EditorOption.lineHeight);
+    const lineCount = editor.getModel()?.getLineCount() || 1;
+    const height = editor.getTopForLineNumber(lineCount + 1) + lineHeight;
+    if (!jsonValue || jsonValue === '') {
+      editorElement.style.height = '0px';
+    } else {
+      editorElement.style.height = `${height}px`;
+    }
+    editor.layout();
+  }, [editor, jsonValue]);
+
+  const loadingState = <EuiLoadingSpinner size="m" />;
+
+  const errorMessageTitle = (
+    <h2>
+      {i18n.translate('discover.sourceViewer.errorMessageTitle', {
+        defaultMessage: 'An Error Occurred',
+      })}
+    </h2>
+  );
+  const errorMessage = (
+    <p>
+      {i18n.translate('discover.sourceViewer.errorMessage', {
+        defaultMessage: 'Could not fetch source at this time. Refresh the tab to try again.',
+      })}
+    </p>
+  );
+  const errorState = (
+    <EuiEmptyPrompt iconType="alert" title={errorMessageTitle} body={errorMessage} />
+  );
+
+  if (reqState === ElasticRequestState.Loading) {
```

Comment:
`loadingState` should also be displayed when `jsonValue === ''`, which is the case at initial rendering

Reply:
this would case would problems with the error state. in the error occurs, the `jsonValue` would never get updated, the error message would never show, and the loading indicator would still show. I think tying loading indicator to request values is enough in this case.

Reply:
You're right, but why not just change the order of checks. first for error, then for loading and for an empty jsonValue? then all should be covered